### PR TITLE
C++: Reuse bounded predicate in TaintedAllocationSize query

### PR DESCRIPTION
**Pull request by paldepind**
Reuse the `bounded` predicate in the `TaintedAllocationSize` query.

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-190/Bounded.qll
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-190/Bounded.qll:30]: Note that it's a pretty common pattern to have three cases to think about (1) a thing holds (2) the thing doesn't hold and (3) we weren't able to determine either.  If we had another predicate, `convertedExprNeverOverflows`, this would not be the inverse of `convertedExprMightOverflow` (due to case 3) and the choice between using `convertedExprNeverOverflows` or `not convertedExprMightOverflow` as we do here would be meaningful.

**Review by geoffw0 — COMMENTED** [review]
Changes LGTM.

We looked at some MRVA results and were happy with the differences.  We should review the DCA results as well before merging.

And since this change does affect results, add a change note.

**Review by geoffw0 — DISMISSED** [review]
Happy with this, when you've finished reviewing a sample of those DCA differences.

**Comment by paldepind** [comment]
I've taken a look at the results in the DCA report. With this PR there are 11 less query results.

One of the disappearing results is in git. The added barrier is [`c & 15`](https://github.com/git/git/blob/477ce5ccd69bda5c44f53e3b230f8a9c54217ed6/builtin/unpack-objects.c#L540) which is correct as that certainly bounds the input.

8 of the disappearing results are in vim. After looking at them, they all go through the same [function with a bitwise `&` barrier](https://github.com/vim/vim/blob/c6ed816761f44da92d8c61f5ea6cb7fdbb45fac5/src/charset.c#L675). Here the mask, `CT_CELL_MASK`, is 7, so this also effectively bounds the value (and even before that the paths are very long (60-120) and kinda questionable). 

The last 2 results are in openjdk. Here the paths are short (10 steps). The [source](https://github.com/jdk-mirror/openjdk8/blob/d15d915ceb0f6a5bfab067a23a9e4a674dac3da8/launcher/parse_manifest.c#L350) reads a file which ends up tainting [an allocation](https://github.com/jdk-mirror/openjdk8/blob/d15d915ceb0f6a5bfab067a23a9e4a674dac3da8/launcher/parse_manifest.c#L88). In between [some bit manipulation](https://github.com/jdk-mirror/openjdk8/blob/d15d915ceb0f6a5bfab067a23a9e4a674dac3da8/launcher/manifest_info.h#L97) goes on, apparently to extract the right bits from a header. I think this could be a true positive for the purpose of the query, but it's clearly deliberate that those bits in the header should end up affecting the allocation.

I think the results seem fine. I think one future approach that could increase precision would be to consider that actual values used in the arithmetic operations (`&`, `/`, `%`, etc.) and only accept barriers where the operation actually reduces the size significantly (such as `% sizeof(char)` but not `/ sizeof(char)`).

**Review by geoffw0 — APPROVED** [review]
Looks great, thanks for addressing this!

**Comment by paldepind** [comment]
Fixed some conflicts after merging the other PR related to `TaintedAllocationSize`. Should be ready for final review now :)

**Review by geoffw0 — APPROVED** [review]
LGTM.

File: cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp
Comment (intrigus-lgtm) [inline cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp:188]: Why is this example "GOOD", but the one from lines 132-136 isn't?
That is, I can't see a real difference here:
https://godbolt.org/z/bPWaoPMYY

Comment (intrigus-lgtm) [reply cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp:188]: I guess my actual question is: how is modulo supposed to bound the input here?

Comment (paldepind) [reply cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp:188]: Thanks for pointing this out @intrigus-lgtm. It's fixed now by #17269.

**Pull request merged**
